Feature/replace to rspec

KensoDev opened the pull request:
Replace Minitest with Rspec.
I tested generating model and some other stuff, all working with rspec.
Main Vendored gem still works with Minitest.

KensoDev committed:
Adding Rspec as a testing framework

* Changed README to reflect the rspec change
* Remove test folder
* Changed templates to work with Rspec
* Mocks with :rr so only added rspec-core and rspec-expectations

brandur commented:
@mfine @pedro Hey guys, any thoughts on this one? I've probably already covered this in depth, but I'm personally not the biggest fan of Minitest (although I will say that it works great for smaller projects). These are probably my top two problems with Minitest:

### Poor command tooling

To run the whole suite, you generally use Rake's `TestTask`; that part is okay. To run a single suite you generally use something like `ruby -Itest path/to/test` or `rake test TEST=path/to/test`; these are okay, but especially with the first `ruby` command, it's not great how you have to initialize a path to get `test_helper` or use `require_relative`.

The worst is the one-off test. Generally you use something like `rake -Itest path/to/test -n /test_pattern/`, but other types of specificity (like line numbers) aren't built in. You also have to know the huge gotcha that a spec like `it "tests for success"` internally gets converted to `test_for_success` and that you have to match on the latter with `-n`.

### Unsanitized backtraces

Minitest dumps big and ugly 50 line traces, including _everything_. RSpec will show you the lines that failed inside your app.

There are fixes for these things that normally ship with extension gems, but you kind of have to question whether you're getting away from the sheer simplicity of Minitest, and if so, whether it might be time to just use a testing framework that supports things that you want to do.

RSpec pushes things the other way and has _way too many_ features, but it's pretty easy to build conventions around the ones that you want to use. RSpec is also slightly slower to start in that it requires more files, but IMO it's unlikely that that will be the weak link in your tests.

&lt;/bikeshed&gt;

Thoughts?

brandur commented:
@KensoDev Awesome man! Thanks for sending this in.

pedro commented:
+1! I usually just start with minitest because it's easy to setup and I like to keep dependencies minimal at first, but sooner or later the lack of better tooling gets really annoying just like you mentioned.

Rspec is big but they did a lot of job to split it into smaller gems. Is it possible to take rspec-mocks out, for instance, considering we're using rr?

brandur commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,46 @@
+App.require!(['lib/models/**/*'])
+
+# pull in test initializers
+Pliny::Utils.require_glob("#{App.root}/test/support/**/*.rb")
```
s/test/spec/

brandur commented on vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb:
```diff
@@ -22,34 +22,34 @@ describe Endpoints::<%= plural_class_name %> do
-    assert_equal 200, last_response.status
-    assert_schema_conform
+    last_response.status.should eq(200)
+    assert_schema_confirm
```
s/confirm/conform/ (and same thing in other specs in this suite)

brandur commented on vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb:
```diff
@@ -22,34 +22,34 @@ describe Endpoints::<%= plural_class_name %> do
+    assert_schema_confirm
   end
 
   it "GET <%= url_path %>/:id" do
```
Doh, I must have screwed this up. Mind changing this to `DELETE`?

brandur commented on vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb:
```diff
@@ -22,34 +22,34 @@ describe Endpoints::<%= plural_class_name %> do
     get "<%= url_path %>/#{@<%= field_name %>.uuid}"
-    assert_equal 200, last_response.status
-    assert_schema_conform
+    last_response.status.should eq(201)
```
s/201/200/ (and same for the `PATCH` and `DELETE`)

brandur commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,46 @@
+require "bundler"
+Bundler.require(:default, :test)
+
+require "minitest/autorun"
```
I think we can remove this require now.

brandur commented:
@KensoDev Also, mind adding `config.mock_framework = :rr`? Thanks!

brandur commented:
@pedro Awesome man, thank-you!

@KensoDev @pedro This is another somewhat arbitrary bikeshed, but like Pedro mentioned above, we technically could just use `rspec-core` and forego its other modules. For example, I'm kind of a fan of the simplicity of Minitest's assertion syntax, which isn't quite so easy to write opaquely compared to RSpec's assertions module. We could do something like `config.expect_with :stdlib`, and just use things like `assert_equal`, `assert_match`, etc. Thoughts?

pedro commented:
+1! Not too concerned with rspec-expectations, but would definitely like to see mocks out.

KensoDev commented:
@brandur @pedro All concerns have been addressed, I updated the PR and rebased with a single commit that describes the work. This way the git log will also make sense.

LMK if you have anything else on your mind.
